### ISSUE TITLE
Reduce Error Logging for Context Cancelled Situations

### ISSUE THREAD

**Issue by bacoboy** [storj/storj#2802]
As a followup to #1743, it has been stated many times also in the forums that the context cancelled errors are not actually errors.  However, they are logged at `ERROR` level with a nasty looking golang stack trace.

For instance:
```
2019-08-15T23:19:42.451Z	INFO	piecestore	upload failed	{"Piece ID": "4RAHYAXYBWJLWNZHV4FPHS6EKD6YIO4E4CJOTGLC3Z7EIO5ABXWA", "SatelliteID": "12EayRS2V1kEsWESU9QMRseFhdxYxKicsiFmxrsLZHeLUtdps3S", "Action": "PUT", "error": "piecestore protocol: rpc error: code = Canceled desc = context canceled", "errorVerbose": "piecestore protocol: rpc error: code = Canceled desc = context canceled\n\tstorj.io/storj/storagenode/piecestore.(*Endpoint).Upload:288\n\tstorj.io/storj/pkg/pb._Piecestore_Upload_Handler:701\n\tstorj.io/storj/pkg/server.(*Server).logOnErrorStreamInterceptor:23\n\tgoogle.golang.org/grpc.(*Server).processStreamingRPC:1134\n\tgoogle.golang.org/grpc.(*Server).handleStream:1185\n\tgoogle.golang.org/grpc.(*Server).serveStreams.func1.1:703"}
2019-08-15T23:19:42.451Z	ERROR	server	piecestore protocol: rpc error: code = Canceled desc = context canceled
	storj.io/storj/storagenode/piecestore.(*Endpoint).Upload:288
	storj.io/storj/pkg/pb._Piecestore_Upload_Handler:701
	storj.io/storj/pkg/server.(*Server).logOnErrorStreamInterceptor:23
	google.golang.org/grpc.(*Server).processStreamingRPC:1134
	google.golang.org/grpc.(*Server).handleStream:1185
	google.golang.org/grpc.(*Server).serveStreams.func1.1:703
```

I'd like to suggest that the second `ERROR` log is redundant and should be removed.  There is no action to be taken and just confuses people for what is a normal operational thing that sometimes happens.

I'd also suggest that the embedded stack trace in the `INFO` message can be removed as it also doesn't provide any useful information for the operator.  `context cancelled` is sufficient in this case (or even change it to `somebody else beat you to it, get a better Internet provider`) ;)

**Comment by ifraixedes** [comment]
@bacoboy Thanks for your feedback.

I agree that the shouldn't be actual errors, but I would log them in debug level.
I also agree on removing the error field from the `INFO` log message.

**Comment by mcumings** [comment]
IMHO, the error field on the INFO log is useful context in order to distinguish this failure mode from others.  I have monitoring which specifically watches for context cancellations and without this I would lose the ability to distinguish this case from other PUT failures.

**Comment by bacoboy** [comment]
@mcumings I agree with you.  But if the error is "handled" and there is nothing to do, then just INFO level that.  Other errors which are not handled and shouting for human interaction would include this level of detail with ERROR level.  So replace this:
```
2019-08-15T23:19:42.451Z	INFO	piecestore	upload failed	{"Piece ID": "4RAHYAXYBWJLWNZHV4FPHS6EKD6YIO4E4CJOTGLC3Z7EIO5ABXWA", "SatelliteID": "12EayRS2V1kEsWESU9QMRseFhdxYxKicsiFmxrsLZHeLUtdps3S", "Action": "PUT", "error": "piecestore protocol: rpc error: code = Canceled desc = context canceled", "errorVerbose": "piecestore protocol: rpc error: code = Canceled desc = context canceled\n\tstorj.io/storj/storagenode/piecestore.(*Endpoint).Upload:288\n\tstorj.io/storj/pkg/pb._Piecestore_Upload_Handler:701\n\tstorj.io/storj/pkg/server.(*Server).logOnErrorStreamInterceptor:23\n\tgoogle.golang.org/grpc.(*Server).processStreamingRPC:1134\n\tgoogle.golang.org/grpc.(*Server).handleStream:1185\n\tgoogle.golang.org/grpc.(*Server).serveStreams.func1.1:703"}
```

With this:
```
2019-08-15T23:19:42.451Z	INFO	piecestore	upload cancelled	{"Piece ID": "4RAHYAXYBWJLWNZHV4FPHS6EKD6YIO4E4CJOTGLC3Z7EIO5ABXWA", "SatelliteID": "12EayRS2V1kEsWESU9QMRseFhdxYxKicsiFmxrsLZHeLUtdps3S", "Action": "PUT"}

```

Its not failed and we don't need the details.  It was `upload cancelled`

**Comment by stefanbenten** [comment]
currently the logging system is more verbose that it should be, simply to have more information in case of a bug for the developer team. The log messages are going to be improved prior to Production Launch.

**Comment by MaxMaximus** [comment]
Hello.

I guess this issue apply in a both ways and "piecestore download failed" with  "Action": "GET", "error": "piecestore: piecestore protocol: context canceled", "errorVerbose": "piecestore: piecestore protocol: context canceled"
Not an actual errors too, but node is "too slow" and other nodes was selected and transfer from this node is canceled?

If so it need new info (instead of "error level" message too).

Only one thing bother me: i see such errors both with upload and download. But for upload it just few % of all upload fails with "context canceled" while for downloads it is about 70-90% range (almost all).
If it due to slow response of node why such huge difference between upload and download? I have a symmetrical internet channel  (50/50 Mbps) on this node. I have read and know about geo location latency but it should affect  upload and download of same node simultaneously.

AND in logs i see a lot of such download fails with same Piece ID. If my node was too slow to react and other nodes already did the work needed - then why satellite keep asking (and keep failing) same Piece of data?
Example of logs on pastebin: https://pastebin.com/nT9MU7ZU

For example:
"Piece ID": "H5LN5CNSFYZAQ2JVXIGS75NJ7YLT4RHYL3V4M4DK46GYO4AMKQCQ"
"Piece ID": "GIK7GBZIZUABHGRU4LOIEZRLCPGQM6CM7SXMIC2M5PWQ2YBFA6NA"

Both pieces was uploaded OK from first try. But later satellite tried to download these pieces many times and all attempts were failed with  "piecestore protocol: context canceled"
Is it still normal way of operations? Or it a different issue - some actual errors happening?

**Comment by ifraixedes** [comment]
> Is it still normal way of operations? Or it a different issue - some actual errors happening?

It's normal.
Your node contains pieces that are valid. The network store redundant pieces of data in the form of erasure codes. Uplink tries to download more pieces than needed to reconstruct the requested data but when it gets enough amount it cancels the rest.

For example, if for reconstructing the data is required 29 pieces and the network has stored 80, uplink requests 80 when downloading the data but when it gets 29, it cancels the rest (51).

**Comment by MaxMaximus** [comment]
I know its  a normal operation to cancel some of transfers by itself. But I doubt it normal to occur in such disproportion between upload and download.
When >90% of downloads canceled while only 1-2% of uploads cancelled on the SAME node with same hardware/sofware and same network connection. 

And about number: 
I thought  it 29 of 35 for downloads from nodes (uplink ask pieces of file from 35 nodes, gets first 29, cancel 6 slowest). 
Isn't it? 
So average "cancel ratio" across whole network should be in ~20% range (while i see ~90%).
And about the same apply to uploads to nodes - satellite send pieces to 95 nodes, 80 fastest get it stored and slowest 15 canceled. 
So average "cancel ratio" for upload and download should be close to each other.
6/35 ~ 17%
15/80 ~ 19%

**Comment by ifraixedes** [comment]
> I know its a normal operation to cancel some of transfers by itself. But I doubt it normal to occur in such disproportion between upload and download.
When >90% of downloads canceled while only 1-2% of uploads cancelled on the SAME node with same hardware/sofware and same network connection.

That makes sense, but it also makes sense to me that we cannot predict what it's uploaded and downloaded and it could be the case that for uploading your node isn't competing with the same nodes as downloading.

For now, we don't have any identified bug which doesn't mean that we don't have any. I'll share your concerns with other team members in case somebody has a better explanation or at least we can be vigilant if this is happening to more storage node operators.

> I thought it 29 of 35 for downloads from nodes (uplink ask pieces of file from 35 nodes, gets first 29, cancel 6 slowest).
Isn't it?

I didn't say the right number, it was an example, that's why I started the sentence :wink: 

> For example...

**Comment by MaxMaximus** [comment]
I read post of few other SNOs on local (rusian) forum witch saw similar issues with download traffic on their nodes while upload traffic was OK. Although it was more mild compared to my node. And in one of cases it is already resolved somehow (from >50% error rate to <5%).
Meantime I have collected some stat for this node. May be give some ideas about causes.
Looks like network delays does not affect this at all. Best download result (~90% succeed/10% fails) are for us-central-1.tardigrade.io satellite where i have relative high network delay (~145 ms).   
And worst results is for europe-west-1.tardigrade.io (100% of GET requests failed)  and satellite.stefan-benten.de (95% GET failed) with low delay (45-70 ms)

```
roundtrip network delay for satellites:
europe-west-1.tardigrade.io - 45 ms
satellite.stefan-benten.de - 70 ms
us-central-1.tardigrade.io - 145 ms
asia-east-1.tardigrade.io - 300 ms

  Satellite: europe-west-1.tardigrade.io        22,40%
Name          Value
----          -----
1. DOWNLOAD   [-------------                 ] 43,1818  19 of 44
2. UPLOAD     [------------------------------] 99,9351  6162 of 6166
3. GET_AUDIT  [------------------------------] 100      19 of 19
4. GET        [                              ] 0        0 of 25
6. PUT        [------------------------------] 99,9349  6144 of 6148
7. PUT_REPAIR [------------------------------] 100      18 of 18
9. EFFECIENCY [------------------------------] 99,533   6181 of 6210


  Satellite: satellite.stefan-benten.de 12,04%
Name          Value
----          -----
1. DOWNLOAD   [--                            ] 5,7862   92 of 1590
2. UPLOAD     [---------------------------   ] 88,8952  1553 of 1747
3. GET_AUDIT  [------------------------------] 100      10 of 10
4. GET        [--                            ] 5,1899   82 of 1580
6. PUT        [--------------------------    ] 88,3172  1459 of 1652
7. PUT_REPAIR [------------------------------] 98,9474  94 of 95
9. EFFECIENCY [---------------               ] 49,2958  1645 of 3337


  Satellite: us-central-1.tardigrade.io        37,17%
Name          Value
----          -----
1. DOWNLOAD   [---------------------------   ] 91,0256  213 of 234
2. UPLOAD     [------------------------------] 99,8312  10054 of 10071
3. GET_AUDIT  [------------------------------] 100      31 of 31
4. GET        [---------------------------   ] 89,6552  182 of 203
6. PUT        [------------------------------] 99,8312  10054 of 10071
9. EFFECIENCY [------------------------------] 99,6312  10267 of 10305

   Satellite: asia-east-1.tardigrade.io      
Name                           Value
----                           -----
1. DOWNLOAD                    [-------------                 ] 43,0769         28 of 65
2. UPLOAD                      [------------------------------] 99,8463         7795 of 7807
3. GET_AUDIT                   [------------------------------] 100             24 of 24
4. GET                         [---                           ] 9,7561          4 of 41
6. PUT                         [------------------------------] 99,8463         7795 of 7807
9. EFFECIENCY                  [------------------------------] 99,3775         7823 of 7872
```

**Comment by ifraixedes** [comment]
@MaxMaximus you input is great, but I'm worried that the topic of the issue is different than you are exposing.

What do you think if you move all this great digging that you did to a new issue and leave this one to just reduce the verbosity in logging due to context canceled situations?

P.S. Meanwhile, I'm going to share your last observations with my Storj mates regarding the different download success on the different satellites.

**Comment by MaxMaximus** [comment]
Yes, i think new issue will be better as it actually different issues. 
I have opened new issue:  [https://github.com/storj/storj/issues/3763](https://github.com/storj/storj/issues/3763)

**Comment by bacoboy** [comment]
Following up on this since I incorrectly said this was still being logged at `ERROR` level in the forums.  It appears this is now `INFO` level as this example shows:

```
2020-04-11T17:09:27.771Z	INFO	piecestore	upload canceled	{"Piece ID": "IS27BL5Q6LMGLXLJXOOGNUZAIJ2HNZRJARXFEULB5H3AHUTVKT7Q", "Satellite ID": "121RTSDpyNZVcEU84Ticf2L1ntiuUimbWgfATz21tuvgk3vzoA6", "Action": "PUT", "error": "context canceled", "errorVerbose": "context canceled\n\tstorj.io/common/pb/pbgrpc.init.0.func3:70\n\tstorj.io/common/rpc/rpcstatus.Wrap:77\n\tstorj.io/storj/storagenode/piecestore.(*Endpoint).doUpload:362\n\tstorj.io/storj/storagenode/piecestore.(*drpcEndpoint).Upload:215\n\tstorj.io/common/pb.DRPCPiecestoreDescription.Method.func1:987\n\tstorj.io/drpc/drpcmux.(*Mux).HandleRPC:107\n\tstorj.io/drpc/drpcserver.(*Server).handleRPC:105\n\tstorj.io/drpc/drpcserver.(*Server).ServeOne:56\n\tstorj.io/drpc/drpcserver.(*Server).Serve.func2:93\n\tstorj.io/drpc/drpcctx.(*Tracker).track:51"}
```

However, I don't think this can be closed as the stack trace is still pretty much noise in this context to the SNO and should be removed.  In theory it would always be the same point in the code that the cancel happens so adds no value.  If there are multiple causes, then that should be another json field.  

A golang stacktrace shouldn't be shown in an INFO status message.  This could be as simple as removing the `errorVerbose` field from this message AND/OR replacing it with something else.

Thoughts?

**Comment by JessicaGreben** [comment]
@bacoboy I will change the log level  of "upload canceled" to `DEBUG` instead of `INFO` and also remove the stack trace. Thanks for the feedback, Its really helpful!

**Comment by bacoboy** [comment]
@JessicaGreben I'm not sure it IS DEBUG level any more than the others, but removing the `errorVerbose` should be enough to remove confusion among new SNOs.

As an SNO I'd move them ALL to DEBUG and create 1 or N (settable) minute/hour summary stat lines at INFO instead -- right now the only way to know what is going on at a summary level over time is to guess from either the webui on port 14002 OR from the terminal dashboard OR shutting down and running a 3rd party tool on the underlying DB.

**Comment by JessicaGreben** [comment]
@bacoboy thanks for the info. In that linked [PR](https://github.com/storj/storj/pull/3857) the verbose error will be removed.

About the summary stats... that is a good idea about the summary log on some interval. I'm curious what specific info would be ideal to display.  I think the original hope was that the terminal dashboard would show all the desired summary info, but sounds like it doesn't have everything you need.  Do you know what specifically you would like to show up there?

**Comment by cdhowie** [comment]
I think leaving them at INFO is fine. In particular, I use zabbix to analyze the logs and look for these lines to graph my upload/download success/failures over time. This is useful information to have in a default configuration.

I worry that having to change the log level to DEBUG is going to cause the logs to grow much more rapidly with a bunch of other internal stuff that isn't relevant to monitoring node health.

**Comment by JessicaGreben** [comment]
@cdhowie makes sense. let me change back to info then.  But I will remove the verbose error stack trace.

**Comment by bacoboy** [comment]
I don't dumpster dive into the DB due to the shutdown requirements, so I'm probably not the best person to ask.  Might be a good Q for the forum users.  I would imagine they'd like something json-y so they could programmatically do log shipping to something else (elastic search, whatever) -- but in terms of what might be important I'm sure you'll get 100 different answers.

Off the top of my head something like +/- delta of data stored (i.e. +1.4MB), consumed bandwidth for that period (in:5MB, out:10kB).  Something in-between the dashboard (mostly static uptime counter) and every request/action...

**Comment by JessicaGreben** [comment]
@bacoboy lol "dumpster dive into the db"

thanks for the details. we will look into it more making the db data more easily accessible.

**Comment by JessicaGreben** [comment]
closing for now, we can re open if more logging needs to be adjusted.

**Comment by ifraixedes** [comment]
> @JessicaGreben I'm not sure it IS DEBUG level any more than the others, but removing the `errorVerbose` should be enough to remove confusion among new SNOs.
> 
> As an SNO I'd move them ALL to DEBUG and create 1 or N (settable) minute/hour summary stat lines at INFO instead -- right now the only way to know what is going on at a summary level over time is to guess from either the webui on port 14002 OR from the terminal dashboard OR shutting down and running a 3rd party tool on the underlying DB.

@bacoboy that's a good idea and we'd love it that you post it in our ideas portal https://ideas.storj.io/

**Comment by MaxMaximus** [comment]
I still see "noise" stack trace debug for context canceled messages after updating to latest release (v. 1.3.3.).
```
2020-04-30T15:38:50.891+0300	INFO	piecestore	download started	{"Piece ID": "3HDJH6IPMPQ6TXEXB2ZX3UUIZWLY4B4I7WK2NZIRPDHH755ZRKFQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "GET"}
2020-04-30T15:38:57.072+0300	INFO	piecestore	download canceled	{"Piece ID": "3HDJH6IPMPQ6TXEXB2ZX3UUIZWLY4B4I7WK2NZIRPDHH755ZRKFQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "GET", "error": "context canceled", "errorVerbose": "context canceled\n\tstorj.io/common/pb/pbgrpc.init.0.func3:70\n\tstorj.io/common/rpc/rpcstatus.Wrap:77\n\tstorj.io/storj/storagenode/piecestore.(*Endpoint).doDownload.func5:653\n\tgolang.org/x/sync/errgroup.(*Group).Go.func1:57"}
2020-04-30T15:43:54.538+0300	INFO	piecestore	download started	{"Piece ID": "BTEZQYDM6ZCW7WZRH37RTSMMI2GLVHWXE66NM6QEZIYUQ7MNETRQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "GET"}
2020-04-30T15:44:02.396+0300	INFO	piecestore	download canceled	{"Piece ID": "BTEZQYDM6ZCW7WZRH37RTSMMI2GLVHWXE66NM6QEZIYUQ7MNETRQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "GET", "error": "context canceled", "errorVerbose": "context canceled\n\tstorj.io/common/pb/pbgrpc.init.0.func3:70\n\tstorj.io/common/rpc/rpcstatus.Wrap:77\n\tstorj.io/storj/storagenode/piecestore.(*Endpoint).doDownload.func5:653\n\tgolang.org/x/sync/errgroup.(*Group).Go.func1:57"}
```
I have thought patch was already merged about 2 weeks ago and should be included in current 1.3.3 ver. Isn’t it?

**Comment by MaxMaximus** [comment]
Update:
Looks like it fixed in ver. 1.3.3 indeed. But only for upload canceled, while still issues for download canceled. Example from a v. 1.3.3 (windows GUI) node:
```
2020-04-30T16:31:22.001+0300	INFO	piecestore	upload started	{"Piece ID": "6RII73UCNI3BDGSBK3PALDSGS5I62WV2I3B6KU4EXWNLQA5JUBXQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "PUT", "Available Space": 156591616}
2020-04-30T16:31:24.790+0300	INFO	piecestore	upload canceled	{"Piece ID": "6RII73UCNI3BDGSBK3PALDSGS5I62WV2I3B6KU4EXWNLQA5JUBXQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "PUT"}
2020-04-30T16:31:32.737+0300	INFO	piecestore	upload started	{"Piece ID": "2KVMVOAZC7PDYOSFW32VWA6IHJWMH3NQHOQ3LANSSRBP2GZ75XFQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "PUT", "Available Space": 154271744}
2020-04-30T16:31:34.231+0300	INFO	piecestore	upload canceled	{"Piece ID": "2KVMVOAZC7PDYOSFW32VWA6IHJWMH3NQHOQ3LANSSRBP2GZ75XFQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "PUT"}
----------cut------------
2020-04-30T16:47:21.485+0300	INFO	piecestore	download started	{"Piece ID": "OZAZNARBKSNIO5PDZYI5LUTFO46T3N6Y7K3N2N2PYKCTLBTI3FIA", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "GET"}
2020-04-30T16:47:24.938+0300	INFO	piecestore	download canceled	{"Piece ID": "OZAZNARBKSNIO5PDZYI5LUTFO46T3N6Y7K3N2N2PYKCTLBTI3FIA", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "GET", "error": "context canceled", "errorVerbose": "context canceled\n\tstorj.io/common/pb/pbgrpc.init.0.func3:70\n\tstorj.io/common/rpc/rpcstatus.Wrap:77\n\tstorj.io/storj/storagenode/piecestore.(*Endpoint).doDownload.func5:653\n\tgolang.org/x/sync/errgroup.(*Group).Go.func1:57"}
2020-04-30T16:47:37.422+0300	INFO	piecestore	download started	{"Piece ID": "ZVP2DBHV4G6V6UGW5SPMVM2PBQO4VDXJXAQ3B772GMBMCCNCBDEQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "GET"}
2020-04-30T16:47:45.203+0300	INFO	piecestore	download canceled	{"Piece ID": "ZVP2DBHV4G6V6UGW5SPMVM2PBQO4VDXJXAQ3B772GMBMCCNCBDEQ", "Satellite ID": "1wFTAgs9DP5RSnCqKV1eLf6N9wtk4EAtmN5DpSxcs8EjT69tGE", "Action": "GET", "error": "context canceled", "errorVerbose": "context canceled\n\tstorj.io/common/pb/pbgrpc.init.0.func3:70\n\tstorj.io/common/rpc/rpcstatus.Wrap:77\n\tstorj.io/storj/storagenode/piecestore.(*Endpoint).doDownload.func5:653\n\tgolang.org/x/sync/errgroup.(*Group).Go.func1:57"}
```

**Comment by bacoboy** [comment]
@JessicaGreben ☝️ -- looks like we need another PR
